### DISCUSSION TITLE
Discussion: String handling pyopenms (python 3)

### DISCUSSION THREAD

**Discussion by oliveralka** [OpenMS/OpenMS#5218] [Code]
There are currently a few issues with string handling in terms of ease of use. 

Currently, due to the integration of both python 2 and python 3 OpenMS::Strings are represented as binary strings (e.g.`b"param1"`). Also functions that expect an OpenMS::String need a binary one.

```
p = Param()
p.setValue("param3", [b"H:+:0.6",b"Na:+:0.2",b"K:+:0.2"], "This is value 3 (StringList)")
```
This also depends on the current implementation of param 
https://github.com/OpenMS/OpenMS/blob/9446dc620ea734179ba4b9e338bf0ae2cd07a84c/src/pyOpenMS/pxds/Param.pxd#L22
```
         get(bytes key, default=None) # wrap-ignore
         __getitem__(bytes key) # wrap-ignore
         __setitem__(bytes key, value) # wrap-ignore
```

On the other hand, functions return unicode string instead of binary even if their type is an OpenMS::String.

```
print("sirius_tmp.getTmpMsFile(): ", sirius_tmp.getTmpMsFile())
/private/var/folders/t7/x82jn_jd09vc_hlq_sqs2jlw0000gn/T/20210310_144125_71754_2.ms
```

This leads to inconsistency when running functions. 
As shown above, sirius_tmp is returned as a regular string instead of a binary one and has to be cast to binary, to be used in the following function (msfile.store()).

```
msfile = SiriusMSFile()
# create temporary filesystem objects
debug_level = 10
sirius_tmp = SiriusTemporaryFileSystemObjects(debug_level)
print("sirius_tmp.getTmpMsFile(): ", sirius_tmp.getTmpMsFile())
print("String(sirius_tmp.getTmpMsFile():", String(sirius_tmp.getTmpMsFile()))

msfile.store(spectra, 
             String(sirius_tmp.getTmpMsFile()), # has to be converted to an "OpenMS::String"
             feature_mapping, 
             feature_only,
             isotope_pattern_iterations, 
             no_mt_info, 
             compound_info)
```

The conclusion after discussion with Sam, Timo and Julianus was, that since python 2 was deprecated in 2020, we could switch to unicode string in every regard. 

I am unsure how that would be done in regard to the current String.pyx [1] and if that has to be changed at other locations in the code as well (such as Param.pxd)

[1] https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/addons/String.pyx

**Comment by jpfeuffer** [comment]
I assume most of our python string handling predates python3 and especially the Cython language_level=3 builds.
For OpenMS::String I think one needs to change the conversion function now to assume that "str" is unicode per default (or do it based on the runtime python version):
https://github.com/OpenMS/OpenMS/blob/55598cf39c73a2092365fbdb51e63defa9b29c6f/src/pyOpenMS/addons/ADD_TO_FIRST.pyx#L51
(But I am not sure if this makes a big difference since the actual type of "str" will be hardcoded at compilation time [of pyopenms] with whatever python version we use).

Also the string handling changes to support python3 unicode strings might just not have been complete:
In things like Param.getitem we still require "bytes".
https://github.com/OpenMS/OpenMS/blob/9446dc620ea734179ba4b9e338bf0ae2cd07a84c/src/pyOpenMS/addons/Param.pyx#L56
If we want to support both pythons, I think we would need to accept the "basestring" type (if that does not work, a raw char array) and encode or not based on the major python version. https://cython.readthedocs.io/en/latest/src/tutorial/strings.html#accepting-strings-from-python-code
If we only want to support python 3, the cython documentation says that cython "basestring" should be enough as a type.

**Reply by timosachsenberg** [reply]
I would be fine with only supporting python3 but I have no clue what we would need to change :)

**Reply by jpfeuffer** [reply]
I am also not sure anymore if what I wrote is correct here.

**Comment by jpfeuffer** [comment]
I thought about it a bit more. I think the possible solutions are:
- use the libcpp_utf8(_output)_string types already supported by autowrap whenever we want unicode support (de/encoding penalty?)
- overwrite ConversionProvider for libcpp_string in OpenMS?
- adapt autowrap to be more flexible (e.g. with parameters what string should translate to). Auto-based on pyVersion? Can it maybe generate two overloads?
